ocaml/gencert/selfcert.ml

```diff
@@ -82,25 +82,19 @@ let call_openssl args =
   [length] bits. *)
 let generate_private_key length =
   let args = ["genrsa"; string_of_int length] in
-  let stdout, _stderr = call_openssl args in
-  stdout
+  try
+    let stdout, _stderr = call_openssl args in
+    Ok stdout
+  with e ->
+    let msg = "generating RSA key failed" in
+    D.error "selfcert.ml: %s" msg ;
+    Debug.log_backtrace e (Backtrace.get e) ;
+    R.error_msg msg
 
 let selfsign' issuer extensions key_length expire =
-  let rsa =
-    try
-      generate_private_key key_length
-      |> Cstruct.of_string
-      |> X509.Private_key.decode_pem
-      |> R.failwith_error_msg
-      |> function
-      | `RSA x ->
-          x
-    with e ->
-      let msg =
-        Printf.sprintf "generating RSA key failed: %s" (Printexc.to_string e)
-      in
-      D.error "%s" msg ; failwith msg
-  in
+  let* s = generate_private_key key_length in
+  let* pem = s |> Cstruct.of_string |> X509.Private_key.decode_pem in
+  let rsa = match pem with `RSA x -> x in
```

Comment:
This results in a runtime exception if it's not an RSA private key, the opposite of what you're trying to do ;)

This is not relevant in the current version of mirage-crypto, but it becomes problematic as soon as we update: https://github.com/mirleft/ocaml-x509/blob/main/lib/private_key.ml#L14

Reply:
perhaps it is easier if I remove this commit

Reply:
I think it's fine to get it all in the same function

Reply:
I don't quite follow how
```
  let* pem = s |> Cstruct.of_string |> X509.Private_key.decode_pem in
  let rsa = match pem with `RSA x -> x in
```
can give an exception ?

I think I will reword the error from `decode_pem` however

Reply:
The match can return an exception:
```
# let privkey = `ESDCE 1;;
val privkey : [> `ESDCE of int ] = `ESDCE 1

# let rsa = match privkey with `RSA x -> x;;
Line 1, characters 10-40:
Warning 8: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
`ESDCE _
Exception: Match_failure ("//toplevel//", 1, 10).
```

Reply:
ah, for some reason I was not seeing this warning. maybe I have an old version of x509. I'll fix this

Reply:
Yeah, this error cannot happening _yet_, but it may be emmited after upgrading to the next version of X509, which has many types of keys.